File: cpp_version/src/services/service_registry.cpp
#include "services/service_registry.h"
#include <algorithm>

void ServiceRegistry::register_service(const std::string& service_name, std::string host, int port) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto& instances = services_[service_name];
    instances.emplace_back(std::move(host), port);
}

void ServiceRegistry::deregister_service(const std::string& service_name, const std::string& host, int port) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = services_.find(service_name);
    if (it != services_.end()) {
        auto& instances = it->second;
        instances.erase(
            std::remove_if(instances.begin(), instances.end(),
                [&](const ServiceInstance& instance) {
                    return instance.host == host && instance.port == port;
                }),
            instances.end());
    }
}

std::vector<ServiceInstance> ServiceRegistry::get_service_instances(const std::string& service_name) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = services_.find(service_name);
    if (it != services_.end()) {
        return it->second;
    }
    return {};
}

File: rust_version/src/services/service_registry.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};

#[derive(Clone, Debug)]
pub struct ServiceInstance {
    pub host: String,
    pub port: u16,
    // 可以添加其他元数据，如健康状态、负载等
}

pub struct ServiceRegistry {
    services: Arc<Mutex<HashMap<String, Vec<ServiceInstance>>>>,
}

impl ServiceRegistry {
    pub fn new() -> Self {
        ServiceRegistry {
            services: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    pub fn register_service(&self, service_name: &str, host: String, port: u16) {
        let mut services = self.services.lock().unwrap();
        let instances = services.entry(service_name.to_string()).or_insert_with(Vec::new);
        instances.push(ServiceInstance { host, port });
    }

    pub fn deregister_service(&self, service_name: &str, host: &str, port: u16) {
        let mut services = self.services.lock().unwrap();
        if let Some(instances) = services.get_mut(service_name) {
            instances.retain(|instance| !(instance.host == host && instance.port == port));
        }
    }

    pub fn get_service_instances(&self, service_name: &str) -> Vec<ServiceInstance> {
        let services = self.services.lock().unwrap();
        services.get(service_name).cloned().unwrap_or_default()
    }
}